src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,302 @@
+  (he : e ∈ G'.edge_set) (hv : v ∈ e) : v ∈ G'.verts :=
+begin
+  refine quotient.ind (λ e he hv, _) e he hv,
+  cases e with v w,
+  simp only [mem_edge_set] at he,
+  cases sym2.mem_iff.mp hv with h h; subst h,
+  { exact G'.edge_vert he, },
+  { exact G'.edge_vert (G'.sym he), },
+end
+
+/-- The `incidence_set` is the set of edges incident to a given vertex. -/
+def incidence_set (G' : subgraph G) (v : V) : set (sym2 V) := {e ∈ G'.edge_set | v ∈ e}
+
+lemma incidence_set_subset_incidence_set (G' : subgraph G) (v : V) :
+  G'.incidence_set v ⊆ G.incidence_set v :=
+λ e h, ⟨G'.edge_set_subset h.1, h.2⟩
+
+lemma incidence_set_subset (G' : subgraph G) (v : V) : G'.incidence_set v ⊆ G'.edge_set :=
+λ _ h, h.1
+
+/-- Give a vertex as an element of the subgraph's vertex type. -/
+@[reducible]
+def vert (G' : subgraph G) (v : V) (h : v ∈ G'.verts) : G'.verts := ⟨v, h⟩
+
+/--
+Given a subgraph, replace the vertex set with an equal set.
+The resulting subgraph is equal (see `copy_eq`).
+-/
+def copy (G' : subgraph G)
+  (V'' : set V) (hV : V'' = G'.verts)
+  (adj' : V → V → Prop) (hadj : adj' = G'.adj) :
+  subgraph G :=
+{ verts := V'',
+  adj := adj',
+  adj_sub := hadj.symm ▸ G'.adj_sub,
+  edge_vert := hV.symm ▸ hadj.symm ▸ G'.edge_vert,
+  sym := hadj.symm ▸ G'.sym }
+
+lemma copy_eq (G' : subgraph G)
+  (V'' : set V) (hV : V'' = G'.verts)
+  (adj' : V → V → Prop) (hadj : adj' = G'.adj) :
+  G'.copy V'' hV adj' hadj = G' :=
+subgraph.ext _ _ hV hadj
+
+/-- The union of two subgraphs. -/
+def union (x y : subgraph G) : subgraph G :=
+{ verts := x.verts ∪ y.verts,
+  adj := λ v w, x.adj v w ∨ y.adj v w,
+  adj_sub := λ v w h, or.cases_on h (λ h, x.adj_sub h) (λ h, y.adj_sub h),
+  edge_vert := λ v w h, or.cases_on h (λ h, or.inl (x.edge_vert h)) (λ h, or.inr (y.edge_vert h)),
+  sym := λ v w h, by rwa [x.adj_comm, y.adj_comm] }
+
+/-- The intersection of two subgraphs. -/
+def inter (x y : subgraph G) : subgraph G :=
+{ verts := x.verts ∩ y.verts,
+  adj := λ v w, x.adj v w ∧ y.adj v w,
+  adj_sub := λ v w h, x.adj_sub h.1,
+  edge_vert := λ v w h, ⟨x.edge_vert h.1, y.edge_vert h.2⟩,
+  sym := λ v w h, by rwa [x.adj_comm, y.adj_comm] }
+
+instance : has_union (subgraph G) := ⟨union⟩
+instance : has_inter (subgraph G) := ⟨inter⟩
+
+/-- The `top` subgraph is `G` as a subgraph of itself. -/
+def top : subgraph G :=
+{ verts := set.univ,
+  adj := G.adj,
+  adj_sub := λ v w h, h,
+  edge_vert := λ v w h, set.mem_univ v,
+  sym := G.sym }
+
+/-- The `bot` subgraph is the subgraph with no vertices or edges. -/
+def bot : subgraph G :=
+{ verts := ∅,
+  adj := λ v w, false,
+  adj_sub := λ v w h, false.rec _ h,
+  edge_vert := λ v w h, false.rec _ h,
+  sym := λ u v h, h }
+
+instance subgraph_inhabited : inhabited (subgraph G) := ⟨bot⟩
+
+/-- The relation that one subgraph is a subgraph of another. -/
+def is_subgraph (x y : subgraph G) : Prop := x.verts ⊆ y.verts ∧ ∀ {v w : V}, x.adj v w → y.adj v w
```

Comment:
hmm, is it possible to write that this is equivalent to `y.coe` having a subgraph isomorphic to `x`? or is that definitionally a huge pain & not useful?

Reply:
I think that would be inconvenient to use, and the obvious interpretation of what you mean would be an inequivalent definition (you want a specific collection of vertices and edges, not just any collection that forms an isomorphic subgraph).  Also, this is the "natural" definition that is in some way induced by the `subgraph` structure.

I invite you to try defining it that way and implementing the `bounded_lattice` instance if this question is out of curiosity.  While evaluating many different definitions of graphs, I've certainly written the instance a number of times myself.

Design-wise, the analogy is that a `simple_graph` is to a `Type` as `subgraph` is to a `set`.  When you have multiple graphs you want to compare, you probably want to make them all be subgraphs of some other graph if you can.  This way they have the same underlying vertex type, which tends to be nicer to deal with.